### ISSUE TITLE
Index Size Error in displaying some PDF files using PDF.js

### ISSUE THREAD

**Issue by kamrapooja** [mozilla/pdf.js#6385]
We are facing an issue while displaying some PDF files using PDF.js in our application. 

PFB other details:
1. PDF files: 
ironman hulkbuster.pdf - https://onedrive.live.com/redir?resid=23F637646F1423!113&authkey=!AMpMZ1CHjLsxivk&ithint=file%2cpdf
Lecota's Transport.pdf - https://onedrive.live.com/redir?resid=23F637646F1423!112&authkey=!AAwrlJacgEFGgeI&ithint=file%2cpdf
2. Location in PDF: Page 1
3. Screenshot: Attached
4. Browser Version Details: Attached
5. Operating System Details: Attached
6. PDF JS version: 1.0.712
7. Console Logs: Attached

The file Lecota's Transport is not even opening in the demo utility provided on Mozilla site when tested under the same environment. Link shared below.

PDF JS Demo: https://mozilla.github.io/pdf.js/web/viewer.html

![application screenshot](https://cloud.githubusercontent.com/assets/9878468/9487567/7acf004c-4bf0-11e5-9b32-9377ebb7095c.jpg)
![browser version](https://cloud.githubusercontent.com/assets/9878468/9487563/7a70bd70-4bf0-11e5-95c3-bb1c5922d757.JPG)
![console logs ironman hulkbuster](https://cloud.githubusercontent.com/assets/9878468/9487564/7a8f32d2-4bf0-11e5-91f5-372b28dec699.JPG)
![console logs leocata transport](https://cloud.githubusercontent.com/assets/9878468/9487565/7a9cd8ce-4bf0-11e5-9b80-a57e741416f9.jpg)
![operating system](https://cloud.githubusercontent.com/assets/9878468/9487566/7ac1c7a6-4bf0-11e5-9ab5-378bd962d20d.JPG)
![pdf js demo screenshot](https://cloud.githubusercontent.com/assets/9878468/9487562/7a3597d6-4bf0-11e5-8ed8-caff6bd3b8ad.jpg)

**Comment by Rob--W** [comment]
PDF.js 1.0.712 is quite old, get the latest version of PDF.js and try again (https://mozilla.github.io/pdf.js/getting_started/#download).

And is his bug specific to Internet Explorer 11? Have you tried other browsers such as Mozilla Firefox or Google Chrome?

**Comment by kamrapooja** [comment]
The demo of PDF.js provided on Mozilla's site is using 1.1.403 release version of PDF.js which is greater than the stable [1.1.215] as well as beta [1.1.366] release available for download.

We checked one of the PDFs [Lecota's Transport.pdf] on the PDF.js demo viewer but it was not able to display the PDF.

Also, this issue is only observed only in this particular version of IE 11 [details shared earlier] and Windows 32 bit operating system for the shared files only. This doesn't occur in any other case.

**Comment by Rob--W** [comment]
I put [PDF.js 1.1.366](https://github.com/mozilla/pdf.js/releases/download/v1.1.366/pdfjs-1.1.366-dist.zip) on a web server, downloaded Lecota's Transport.pdf and also put it on the webserver and then opened the page on the following platforms:

- IE 11.0.10011.0 (Windows 10)
- IE 11.0.9600.16663 (Windows 8.1)
- IE 11.0.9600.17633 (Windows 7).

In all cases, the PDF was successfully displayed.

What are the values of the `paintWidth`, `paintHeight`, `height` and `width` variables when the error occurs?

**Comment by kamrapooja** [comment]
We have only observed this issue only in the specified environment.

Please try to test it in the same environment, as the issue is not observed in any other environment.

PFA the required parameter details with expanded view of "imgToPaint" parameter.

![value details1](https://cloud.githubusercontent.com/assets/9878468/9518062/9a1fa44c-4cd1-11e5-8086-e80f3ddaa835.jpg)
![value details2](https://cloud.githubusercontent.com/assets/9878468/9518061/9a1b8d1c-4cd1-11e5-93d1-a556405dac41.jpg)
![value details3](https://cloud.githubusercontent.com/assets/9878468/9518064/9a289548-4cd1-11e5-8948-226fbced97ee.jpg)
![value details4](https://cloud.githubusercontent.com/assets/9878468/9518063/9a24a3fc-4cd1-11e5-87b7-341325e4ed21.jpg)

**Comment by Rob--W** [comment]
@kamrapooja I don't have access to an identical environment. Could you provide remote desktop access (e.g. via TeamViewer) to allow me to remotely debug the issue?

**Comment by kamrapooja** [comment]
Please share suitable timings in GMT, if any. Please share expected time duration as it's a client machine.

**Comment by Rob--W** [comment]
@kamrapooja Could you follow up by mail? See my profile for contact details.

**Comment by Rob--W** [comment]
I've found the root cause of the issue, and created a minimal test case at https://robwu.nl/pdfjs/6385/reduced/. This problem only occured on the systems provided by @kamrapooja, and it is not caused by the specific version of IE (I've set up an identical version of IE, and couldn't reproduce the bug).

The issue is that on that system, the canvas max size is 2048x2048 when HWA is enabled. This can easily be shown by opening the viewer demo (https://mozilla.github.io/pdf.js/web/viewer.html) and changing the zoom to 400%. The right side of the canvas remains blank.
**Work-around**: Disable HWA (IE: Internet Options -> Advanced -> Use software rendering instead of GPU rendering).

I think that we should somehow detect the canvas size limitations, and make sure that the canvas's size fits within that constraint (and otherwise create multiple canvases).
